commons-lang3: new port

chrstphrchvz opened the pull request:
Based on submission by @someuser12: https://trac.macports.org/ticket/58756 (I have also fixed whitespace and livecheck, and removed sha1 and md5 checksums since upstream only provides sha512 and PGP signature.)

#### Description
To be used with a future `pdftk-java` port.
<!-- Note: it is best make pull requests from a branch rather than from master -->
###### ~~Tested on~~
<!-- Generate version information with this command in shell:
    echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)"; echo "Xcode $(xcodebuild -version | awk '{print $NF}' | tr '\n' ' ')"
-->
~~macOS 10.x~~
~~Xcode 8.x~~

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?
<!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

mf2k commented:
Error is:

`[ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.1.0:javadoc (default-cli) on project commons-lang3: An error has occurred in Javadoc report generation: Unable to find javadoc command: The environment variable JAVA_HOME is not correctly set. -> [Help 1]
`

Try adding the java portgroup to fix that issue.

chrstphrchvz commented:
> Error is:
> 
> `[ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.1.0:javadoc (default-cli) on project commons-lang3: An error has occurred in Javadoc report generation: Unable to find javadoc command: The environment variable JAVA_HOME is not correctly set. -> [Help 1] `
> 
> Try adding the java portgroup to fix that issue.

Thanks for finding the error. Yes, this port needs to specify its dependency on Java 8+ (see https://commons.apache.org/proper/commons-lang/release-notes/RELEASE-NOTES-3.9.txt), and the Java portgroup sounds like the right way to do so. (The `commons-lang` 2.x port this submission is based on only uses `depends_lib bin:java:kaffe`, so I might also update it to also use the Java portgroup; see #4927.)

mf2k commented:
Progress! New error below. Unfortunately, I have no idea what do about it. 

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  04:52 min
[INFO] Finished at: 2019-08-07T15:30:43Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.1.0:javadoc (default-cli) on project commons-lang3: An error has occurred in Javadoc report generation: 
[ERROR] Exit code: 1 - javadoc: error - The code being documented uses modules but the packages defined in https://docs.oracle.com/javase/7/docs/api/ are in the unnamed module.
[ERROR] javadoc: error - The code being documented uses modules but the packages defined in https://docs.oracle.com/javaee/6/api/ are in the unnamed module.
[ERROR] javadoc: error - The code being documented uses modules but the packages defined in https://docs.oracle.com/javase/8/docs/api/ are in the unnamed module.
[ERROR] 
[ERROR] Command line was: /Library/Java/JavaVirtualMachines/zulu-12.jdk/Contents/Home/bin/javadoc @options @packages
[ERROR] 
[ERROR] Refer to the generated Javadoc files in '/opt/local/var/macports/build/_Users_vsts_agent_2.155.1_work_1_s_java_commons-lang3/commons-lang3/work/commons-lang3-3.9-src/target/site/apidocs' dir.
[ERROR] 
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

```

chrstphrchvz commented:
> Port commons-lang3 **fail** on xcode10.2. [Log](https://paste.macports.org/157d5f61dc0c)

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.1.0:javadoc (default-cli) on project commons-lang3: An error has occurred in Javadoc report generation: 
[ERROR] Exit code: 1 - javadoc: error - The code being documented uses modules but the packages defined in https://docs.oracle.com/javase/7/docs/api/ are in the unnamed module.
[ERROR] javadoc: error - The code being documented uses modules but the packages defined in https://docs.oracle.com/javaee/6/api/ are in the unnamed module.
[ERROR] javadoc: error - The code being documented uses modules but the packages defined in https://docs.oracle.com/javase/8/docs/api/ are in the unnamed module.
```

This is a known OpenJDK issue that should already be fixed in updated JDK versions: https://bugs.openjdk.java.net/browse/JDK-8212233

> Port commons-lang3 **fail** on xcode9.4. [Log](https://paste.macports.org/2e4776f98cd3)

```
[ERROR] javadoc: error - An internal exception has occurred. 
[ERROR] 	(com.sun.tools.javac.code.ClassFinder$BadClassFile: bad class file: /opt/local/var/macports/build/_Users_travis_build_macports_macports-ports_java_commons-lang3/commons-lang3/work/commons-lang3-3.9-src/target/commons-lang3-3.9.jar(/org/apache/commons/lang3/ArrayUtils$1.class)
[ERROR]   undeclared type variable: T
[ERROR]   Please remove or make sure it appears in the correct subdirectory of the classpath.)
[ERROR] Please file a bug against the javadoc tool via the Java bug reporting page
[ERROR] (http://bugreport.java.com) after checking the Bug Database (http://bugs.java.com)
[ERROR] for duplicates. Include error messages and the following diagnostic in your report. Thank you.
```

This is a known javadoc issue that doesn't appear to be fixed: https://bugs.openjdk.java.net/browse/JDK-8187422

Should this port attempt to work around these issues? I haven't yet found whether the upstream commons-lang3 project is aware of these issues.

mf2k commented:
> Should this port attempt to work around these issues? I haven't yet found whether the upstream commons-lang3 project is aware of these issues.

Yes definitely, we can't update the Portfile and have it not actually build. If you can work around it, then please do. Otherwise, we will have to wait for the fixes in OpenJDK to propagate.

chrstphrchvz commented:
One other approach is to use the binary distfile (at least by default) rather than compiling it. Is that more desirable than trying to fix the source build? If not, then I will probably look some more at other package managers (that build this from source) to see if they have already addressed these issues.

chrstphrchvz commented:
I have set the required JDK version to exactly 8. This is probably the best I can do for now as far as working around the issues with later JDK versions.

I have also disabled running tests during the build, since the tests for the StopWatch component tend to fail on CI bots; looking at the source of those tests, I think they are intended for running on an idle system rather than a busy CI VM, and can fail if there's a delay of more than a few tenths of a second or so. The tests probably shouldn't be run during the normal build anyways; a future task would be to allow running them with `port test`.

chrstphrchvz commented:
Merged as part of #5111.
